src/IO/WriteBuffer.h

```diff
@@ -60,6 +60,11 @@ class WriteBuffer : public BufferBase
       */
     virtual ~WriteBuffer() {}
 
+    /**
+     * some buffers (kafka / rabbit) split the rows internally, so we can push there formats without
+     * framing / delimiters (like ProtobufSingle)
+     */
+    virtual bool producesIsolatedRows() { return false; }
```

Comment:
It looks really out of place here. `WriteBuffer` is just an output stream that doesn't know anything about rows or Kafka or whatever.

Reply:
> is just an output byte stream that doesn't know anything 

Kafka / Rabbit has very special Buffer implemenation (a bit unnatural).

> It looks really out of place here.

I'm also not very happy about that, but the fact is that was the simplest / the cleanest way i was able to find. Can you propose some better options?

Reply:
I don't think we need this special function.

And the format should always be the same, the `Protobuf` format must not silently turn into the `ProtobufSingle` format in some cases. If we want to prohibit an user from using `Protobuf` format because `WriteBufferToKafkaProducer` doesn't work well with the `Protobuf` format then I think the factory of ProtobufRowOutputFormat could throw an exception if a `WriteBuffer` passed in here is of the type `WriteBufferToKafkaProducer`.

Reply:
It's not about that. 

Protobuf - can be written (any number of rows) to any buffer. 
ProtobufSingle - for any buffer (like select from clickhouse-client) we allow to produce single row only, for Kafka/Rabbit we allow any number of rows in that format.

Reply:
> So the result i wanted to archive = exception when ProtobufSingle is used with several rows out of kafka / rabbit.

This logic has no direct connection neither to `WriteBuffer` nor to `ProtobufWriter`. So I think the better place to perform this check is where `WriteBuffer` and `ProtobufWriter` are connection together, i.e. the class `ProtobufRowOutputFormat. The function `ProtobufRowOutputFormat::write()` can do this check before starting the first message.